src/memote/network.jl

```diff
@@ -0,0 +1,184 @@
+    )
+    blocked_rids = String[]
+    for rid in reactions(stdmodel)
+        isapprox(abs(mins[rid][rid]), 0.0) &&
+            isapprox(abs(maxs[rid][rid]), 0.0) &&
+            push!(blocked_rids, rid)
+    end
+    return blocked_rids
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Helper function to find orphan or deadend metabolites. Specify `consumed=true`
+to consider orphan metabolites or `false` to consider deadend metabolites. Set
+`complete_medium=true` to open all boundary reactions to simulate a complete
+medium.
+"""
+function _find_orphan_or_deadend_metabolites(model; consumed = true)
+    mids = metabolites(model)
+    mets = String[]
+    S = stoichiometry(model)
+    lbs, ubs = bounds(model)
+    for idx in axes(S, 1)
+        rids, vals = findnz(S[idx, :])
+        if length(vals) == 1 && (consumed ? first(vals) < 0 : first(vals) > 0)
+            ridx = first(rids)
+            rid = reactions(model)[ridx]
+            met = mids[idx]
+            v = reaction_stoichiometry(model, rid)[met]
+            # check if reaction can actually make or consume the metabolite
+            lbs[ridx] < 0 < ubs[ridx] && continue # ignore reversible reactions
+            consumed && lbs[ridx] * v <= 0 && push!(mets, met)
+            !consumed && ubs[ridx] * v >= 0 && push!(mets, met)
+        end
+    end
+    return mets
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Find all metabolites that can only (excludes reversible reactions) be consumed
+in the `model` by inspecting the stoichiometric matrix.
+"""
+find_orphan_metabolites(model) = _find_orphan_or_deadend_metabolites(model, consumed = true)
+
+"""
+$(TYPEDSIGNATURES)
+
+Find all metabolites that can only (excludes reversible reactions) be produced
+in the `model` by inspecting the stoichiometric matrix.
+"""
+find_deadend_metabolites(model) =
+    _find_orphan_or_deadend_metabolites(model, consumed = false)
+
+"""
+$(TYPEDSIGNATURES)
+
+Find all reactions that participate in stoichiometrically balanced cycles by
+closing all boundary reactions and running fva on the resultant model.
+"""
+function find_cycle_reactions(model, optimizer; config = memote_config)
+    stdmodel = convert(StandardModel, model)
+    for rid in reactions(stdmodel)
+        if is_boundary(stdmodel, rid)
+            change_bound!(stdmodel, rid, lower = 0, upper = 0)
+        else # remove fixed constraints
+            if stdmodel.reactions[rid].lb < 0
+                stdmodel.reactions[rid].lb = -1000.0
+            else
+                stdmodel.reactions[rid].lb = 0
+            end
+            if stdmodel.reactions[rid].ub > 0
+                stdmodel.reactions[rid].ub = 1000.0
+            else
+                stdmodel.reactions[rid].ub = 0
+            end
+        end
+    end
+    cycle_reactions = Set(String[])
+    for rid in filter(x -> !is_boundary(model, x), reactions(model))
+        for sense in [MIN_SENSE, MAX_SENSE]
+            mu = solved_objective_value(
```

Comment:
screeeeeeeeeeeeen! :grin:

Reply:
will make parallel in another pr, other functions can also be distributed